core/src/main/scala/kafka/log/Log.scala

```diff
@@ -677,6 +672,58 @@ class Log(val dir: File,
   }
 
   /**
+    * Find any segments that match the user-supplied predicate UNLESS it is the final segment
+    * and it is empty (since we would just end up re-creating it)
+    * @param predicate A function that takes in a single log segment and returns true iff it is deletable
+    * @return the segments ready to be deleted
+    */
+  private def deletableSegments(predicate: LogSegment => Boolean) = {
+    val lastEntry = segments.lastEntry
+    if (lastEntry == null) Seq.empty
+    else logSegments.takeWhile(s => predicate(s) && (s.baseOffset != lastEntry.getValue.baseOffset || s.size > 0))
+  }
+
+  /**
+    * Delete any log segments that have either expired due to time based retention
+    * or because the log size is > retentionSize
+    */
+  def deleteOldSegments(): Int = {
+    if (!config.delete) return 0
+    retentionMsExpiredSegments(deleteOldSegments) + retentionSizeBreachedSegments(deleteOldSegments)
+  }
+
+  private def retentionMsExpiredSegments(runWithPredicate: (LogSegment => Boolean) => Int) : Int = {
```

Comment:
The runWithPredicate logic works, but takes a bit of time to understand. Since deleting a segment is cheaper, I am wondering if it would be simpler if we get rid of shouldDeleteSegments() and just keep deleteOldSegments(). In LogCleaner, instead of first finding logs with segments ready to be deleted, we just get all logs configured with compact and delete and call deleteOldSegments() directly.